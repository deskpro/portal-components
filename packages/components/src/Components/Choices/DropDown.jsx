import React from 'react';
import PropTypes from 'prop-types';
import { getIn } from 'formik';
import { css } from 'emotion';
import ReactSelect, { components } from 'react-select';
import AsyncSelect from 'react-select/lib/Async';
import classNames from 'classnames';
import Field from '../Field';


export const SelectContainer = ({ children, ...props }) => (
  <components.SelectContainer
    {...props}
    className={classNames('react-select', { 'react-select__is-focused': props.isFocused })}
  >
    {children}
  </components.SelectContainer>
);

SelectContainer.propTypes = components.SelectContainer.propTypes;

const Option = (props) => {
  const {
    className,
    cx,
    data,
    getStyles,
    innerRef,
    label,
    isDisabled,
    isFocused,
    isSelected,
    innerProps
  } = props;
  if (data.children && data.children.length > 1) {
    return (
      <div
        ref={innerRef}
        className={cx(
          css(getStyles('option', props)),
          {
            option:                true,
            'option--is-disabled': isDisabled,
            'option--is-focused':  isFocused,
            'option--is-selected': isSelected,
          },
          className
        )}
        {...innerProps}
      >
        {label}
        <span className="react-select__option--arrow" />
      </div>
    );
  } return <components.Option {...props} />;
};

Option.propTypes = components.Option.propTypes;

export class DropDownInput extends React.Component {
  static propType = {
    form:       PropTypes.object,
    dataSource: PropTypes.shape({
      getOptions: PropTypes.oneOfType([PropTypes.func, PropTypes.array]).isRequired,
    }).isRequired,
    handleChange: PropTypes.func,
    isClearable:  PropTypes.bool,
  };

  constructor(props) {
    super(props);
    this.state = {
      value:      null,
      menuIsOpen: false,
      options:    props.dataSource.getOptions,
    };
  }

  onBlur = (form) => {
    const { name } = this.props;
    this.props.onBlur();
    this.setState({
      menuIsOpen: false
    });
    if (form) {
      form.setFieldTouched(name, true);
    }
  };

  onFocus = () => {
    this.props.onFocus();
    this.setState({
      menuIsOpen: true
    });
  };

  handleChange = (form, value) => {
    if (value && value.children && value.children.length) {
      const { children } = value;
      const options = [{
        label:   'Back',
        value:   'select-back',
        parents: this.state.options,
      }].concat(children);
      this.setState({
        value: null,
        options,
      });
      return false;
    } else if (value && value.parents) {
      this.setState({
        value:   null,
        options: value.parents
      });
      return false;
    }
    const { name, handleChange } = this.props;
    this.setState({
      value,
      menuIsOpen: false,
    });
    this.select.blur();
    const newValue = value ? value.value : null;
    form.setFieldValue(name, newValue);
    handleChange(newValue);
    return true;
  };

  loadOptions = (form, inputValue) => {
    const { dataSource, name } = this.props;
    return dataSource.getOptions(inputValue).then((options) => {
      const value = options.find(o => o.value === getIn(form.values, name));
      this.setState({
        value
      });
      return options;
    });
  };

  render() {
    const {
      name, dataSource, isClearable, form, ...props
    } = this.props;
    if (Array.isArray(dataSource.getOptions)) {
      return (
        <ReactSelect
          ref={(c) => { this.select = c; }}
          value={dataSource.getOptions.find(o => o.value === getIn(form.values, name))}
          name={name}
          isClearable={isClearable}
          components={{ SelectContainer, Option }}
          menuIsOpen={this.state.menuIsOpen}
          onChange={value => this.handleChange(form, value)}
          options={this.state.options}
          closeMenuOnSelect={this.closeMenuOnSelect}
          classNamePrefix="react-select"
          {...props}
          onFocus={this.onFocus}
          onBlur={() => this.onBlur(form)}
        />
      );
    }
    return (
      <AsyncSelect
        ref={(c) => { this.select = c; }}
        value={this.state.value}
        name={name}
        isClearable={isClearable}
        onChange={value => this.handleChange(form, value)}
        onBlur={() => form.setFieldTouched(name, true)}
        defaultOptions
        cacheOptions
        components={{ SelectContainer, Option }}
        loadOptions={inputValue => this.loadOptions(form, inputValue)}
        classNamePrefix="react-select"
        {...props}
      />
    );
  }
}

class DropDown extends Field {
  renderField(form) {
    return (
      <DropDownInput
        form={form}
        {...this.props}
      />
    );
  }
}

DropDown.propTypes = {
  ...Field.propTypes,
  dataSource: PropTypes.shape({
    getOptions: PropTypes.oneOfType([PropTypes.func, PropTypes.array]).isRequired,
  }).isRequired,
  handleChange: PropTypes.func,
  isClearable:  PropTypes.bool,
};

DropDown.defaultProps = {
  handleChange() {},
  isClearable: true,
  onBlur() {},
  onFocus() {},
};

export default DropDown;
