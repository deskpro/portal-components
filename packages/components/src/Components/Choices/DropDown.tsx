import * as React from 'react';
import { getIn } from 'formik';
import { css } from '@emotion/css'
import ReactSelect, { components } from 'react-select';
import AsyncSelect from 'react-select/async';
import { deepMerge } from '@deskpro/js-utils/dist/objects';
import classNames from 'classnames';
import Field, { FieldProps, FieldState } from '../Field';
import { ContainerProps } from "react-select/dist/declarations/src/components/containers";
import { DropdownIndicatorProps as ReactSelectDropdownIndicatorProps } from "react-select/dist/declarations/src/components/indicators";
import {
  OptionProps as ReactSelectOptionProps
} from "react-select/dist/declarations/src/components/Option";
import type { Option } from "../../types/options";
import { DataSource } from "../../types/dataSource";

const SelectContainer = ({ children, ...props }: ContainerProps) => (
  <components.SelectContainer
    {...props}
    className={classNames(
      'react-select',
      { 'react-select__is-focused': props.isFocused, 'react-select-multi': props.isMulti }
    )}
  >
    {children}
  </components.SelectContainer>
);

interface DropdownIndicatorProps extends ReactSelectDropdownIndicatorProps {
  closeMenu: () => void;
}

const DropdownIndicator = ({ closeMenu, ...props }: DropdownIndicatorProps) => (
  <components.DropdownIndicator {...props}>
    <span className="dp-react-select__dropdown-indicator" onClick={closeMenu} onKeyDown={() => true} />
  </components.DropdownIndicator>
);

interface OptionProps extends ReactSelectOptionProps {
  data: {
    value: string;
    children: object[];
  };
}

const Option = (props: OptionProps) => {
  const {
    className,
    cx,
    data,
    getStyles,
    innerRef,
    label,
    isDisabled,
    isFocused,
    isSelected,
    innerProps
  } = props;
  if (data.children && data.children.length > 0) {
    return (
      <div
        ref={innerRef}
        className={classNames(
          cx(
            {
              option:                true,
              'option--is-disabled': isDisabled,
              'option--is-focused':  isFocused,
              'option--is-selected': isSelected,
            },
            className
          ),
          css(getStyles('option', props))
        )}
        {...innerProps}
      >
        {label}
        <span className="react-select__option--arrow" />
      </div>
    );
  }
  if (data.value === 'select-back') {
    return (
      <div
        ref={innerRef}
        className={classNames(
          cx(
          {
            option:                true,
            'option--is-disabled': isDisabled,
            'option--is-focused':  isFocused,
            'option--is-selected': isSelected,
            'option--is-back':     true,
          },
          className),
          css(getStyles('option', props))
        )}
        {...innerProps}
      >
        <span className="react-select__option--back-arrow" />
        {label}
      </div>
    );
  }
  return <components.Option {...props} />;
};

const I18N = {
  back:   'Back',
  select: 'Select',
};

interface DropDownInputProps extends FieldProps {
  dataSource:    DataSource;
  onChange:      (value: string | null) => void;
  isClearable?:  boolean;
  isSearchable?: boolean;
  value:         number | string;
  closeOnBlur?:  boolean;
}

interface DropDownInputState extends FieldState {
  value?: {
    value: string;
    children: string[];
  };
  menuIsOpen: boolean;
  options: Option[];
}

export class DropDownInput extends React.Component<DropDownInputProps, DropDownInputState> {
  static defaultProps = {
    i18n: {},
    onBlur() {},
    onFocus() {},
  };
  private i18n: any;
  private readonly select: React.RefObject<any>;

  constructor(props) {
    super(props);

    this.i18n = deepMerge(I18N, props.i18n);
    this.select = React.createRef();

    this.state = {
      value:      null,
      menuIsOpen: false,
      options:    props.dataSource.getOptions,
    };
  }

  componentDidMount() {
    this.setOptions();
  }

  componentDidUpdate(prevProps, prevState) {
    this.setOptions(prevState);
  }

  onBlur = () => {
    this.props.onBlur();
    if (this.props.closeOnBlur) {
      this.setState({ menuIsOpen: false });
    }
  };

  onFocus = () => {
    this.props.onFocus();
    this.setState({
      menuIsOpen: true
    });
  };

  onChange = (value) => {
    if (value && value.children && value.children.length > 0) {
      const { children } = value;
      const options: Option[] = [{
        label:   this.i18n.back,
        value:   'select-back',
        parents: this.state.options,
      }].concat(children);
      this.setState({
        value: null,
        options,
      });
      this.props.onChange(null);
      return false;
    } if (value && value.parents && value.parents.length > 0) {
      this.setState({
        value:   null,
        options: value.parents
      });
      this.props.onChange(null);
      return false;
    }
    this.setState({
      value,
      menuIsOpen: false,
    });
    this.select.current.inputRef.blur();
    const newValue = value ? value.value : null;
    this.props.onChange(newValue);
    return true;
  };

  closeMenu = () => {
    if (!this.props.closeOnBlur) {
      this.setState({ menuIsOpen: false });
      this.select.current.inputRef.blur();
    }
  };

  loadOptions = (inputValue) => {
    const { dataSource } = this.props;
    const propValue = this.props.value;
    if (Array.isArray(dataSource.getOptions)) {
      return dataSource.getOptions;
    }
    return dataSource.getOptions(inputValue).then((options) => {
      const value = options.find(o => o.value === propValue);
      this.setState({
        value
      });
      return options;
    });
  };

  setOptions = (prevState = null) => {
    const { value: stateValue } = this.state;
    const { value: propValue, dataSource } = this.props;

    if (Array.isArray(dataSource.getOptions) && propValue && (!stateValue || stateValue.value !== propValue)) {
      const updateOptions = (options, parentOptions?: Option[]) => {
        if (options.map(option => option.value).indexOf(propValue) !== -1) {
          if (prevState && prevState.options
            && JSON.stringify(options) === JSON.stringify(prevState.options.filter(o => o.value !== 'select-back'))
          ) {
            return;
          }

          const newValue = options.find(o => o.value === propValue);

          if (parentOptions) {
            this.setState({
              value:   newValue,
              options: [{
                label:   this.i18n.back,
                value:   'select-back',
                parents: parentOptions,
              }].concat(options)
            });
          } else {
            this.setState({ value: newValue, options });
          }
        } else {
          options.forEach((childOption) => {
            if (childOption.children) {
              updateOptions(childOption.children, options);
            }
          });
        }
      };

      updateOptions(dataSource.getOptions);
    }
  };

  render() {
    const {
      name, dataSource, isClearable, isSearchable, value, ...props
    } = this.props;
    const { options } = this.state;
    if (Array.isArray(dataSource.getOptions)) {
      return (
        <ReactSelect
          styles={{ menuPortal: base => ({ ...base, position: 'relative' }) }}
          ref={this.select}
          value={options.find(o => o.value === value) || null}
          name={name}
          isClearable={isClearable}
          isSearchable={isSearchable}
          components={{
            SelectContainer,
            Option,
            DropdownIndicator: dropdownProps => (
              <DropdownIndicator closeMenu={this.closeMenu} {...dropdownProps} />
            )
          }}
          menuPlacement="auto"
          menuIsOpen={this.state.menuIsOpen}
          options={options}
          closeMenuOnScroll={false}
          classNamePrefix="react-select"
          placeholder={this.i18n.select}
          {...props}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          onChange={this.onChange}
        />
      );
    }
    return (
      <AsyncSelect
        ref={this.select}
        value={this.state.value}
        name={name}
        isClearable={isClearable}
        isSearchable={isSearchable}
        defaultOptions
        cacheOptions
        components={{
          SelectContainer,
          Option,
          DropdownIndicator: dropdownProps => <DropdownIndicator closeMenu={this.closeMenu} {...dropdownProps} />
        }}
        loadOptions={inputValue => this.loadOptions(inputValue)}
        classNamePrefix="react-select"
        {...props}
        onChange={this.onChange}
        onBlur={this.onBlur}
        onFocus={this.onFocus}
      />
    );
  }
}

interface DropDownProps extends FieldProps {
  dataSource:    DataSource;
  form?:         object;
  handleChange?: (value: any) => void;
  closeOnBlur?:  boolean;
  onBlur?:       () => void;
  onFocus?:      () => void;
  isClearable?:  boolean;
  isSearchable?: boolean;
}

class DropDown extends Field<DropDownProps> {
  static defaultProps = {
    ...Field.defaultProps,
    handleChange() {},
    isClearable:  false,
    isSearchable: true,
    closeOnBlur:  true,
    onBlur() {},
    onFocus() {},
  }

  onBlur = (form) => {
    this.props.onBlur();
    if (this.props.closeOnBlur) {
      const { name } = this.props;
      if (form) {
        form.setFieldTouched(name, true);
      }
    }
  };

  onChange = (form, value) => {
    const { name, handleChange } = this.props;
    const actualVale = typeof value === 'object' && value !== null ? value.value : value;
    const newValue = actualVale || null;
    form.setFieldValue(name, newValue);
    handleChange(newValue);
  };


  renderField({ form }) {
    return (
      <DropDownInput
        onBlur={() => this.onBlur(form)}
        onChange={value => this.onChange(form, value)}
        value={getIn(form.values, this.props.name)}
        {...this.props}
      />
    );
  }
}

export {
  DropdownIndicator,
  OptionProps,
  SelectContainer,
};
export default DropDown;
