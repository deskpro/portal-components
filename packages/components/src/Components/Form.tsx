import * as React from 'react';
import { Form as FormikForm } from 'formik';
import classNames from 'classnames';

interface IProps {
  className?: string;
  showHover?: boolean;
  children:   React.ReactNode;
}

const Form = ({
  className = '',
  showHover = true,
  ...props
}: IProps) => (
  <FormikForm
    className={classNames('dp-pc_form', className, { 'show-hover': showHover })}
    {...props}
  />);



export default Form;
