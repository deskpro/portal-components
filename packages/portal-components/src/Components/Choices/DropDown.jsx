import React from 'react';
import PropTypes from 'prop-types';
import { getIn } from 'formik';
import ReactSelect from 'react-select';

import Field from '../Field';

class Select extends Field {
  handleBlur = () => {
    // this is going to call setFieldTouched and manually update touched.topcis
    this.props.onBlur('topics', true);
  };

  renderField = (form) => {
    const { name, options, ...props } = this.props;
    return (
      <ReactSelect
        value={getIn(form.values, name)}
        name={name}
        onChange={value => form.setFieldValue(name, value)}
        onBlur={value => form.setFieldTouched(name, value)}
        options={options}
        {...props}
      />
    );
  }
}

Select.propTypes = Object.assign(Field.propTypes, {
  options: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
    value: PropTypes.any.isRequired
  }))
});

export default Select;
