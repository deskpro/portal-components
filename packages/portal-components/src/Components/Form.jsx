import React from 'react';
import { Form as FormikForm } from 'formik';

const Form = ({
  ...props
}) => <FormikForm {...props} />;

export default Form;
