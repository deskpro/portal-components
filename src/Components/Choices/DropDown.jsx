import React from 'react';
import PropTypes from 'prop-types';
import { getIn } from 'formik';
import ReactSelect, { components } from 'react-select';
import AsyncSelect from 'react-select/lib/Async';
import classNames from 'classnames';
import Field from '../Field';


const SelectContainer = ({ children, ...props }) => (
  <components.SelectContainer {...props} className={classNames({ 'react-select__is-focused': props.isFocused })}>
    {children}
  </components.SelectContainer>
);

SelectContainer.propTypes = components.SelectContainer.propTypes;

class DropDown extends Field {
  constructor(props) {
    super(props);
    this.state = {
      value: null
    };
  }

  handleChange = (form, value) => {
    const { name, handleChange } = this.props;
    this.setState({
      value
    });
    const newValue = value ? value.value : null;
    form.setFieldValue(name, newValue);
    handleChange(newValue);
  };

  loadOptions = (form, inputValue) => {
    const { dataSource, name } = this.props;
    return dataSource.getOptions(inputValue).then((options) => {
      const value = options.find(o => o.value === getIn(form.values, name));
      this.setState({
        value
      });
      return options;
    });
  };

  renderField = (form) => {
    const {
      name, dataSource, isClearable, ...props
    } = this.props;
    if (Array.isArray(dataSource.getOptions)) {
      return (
        <ReactSelect
          value={dataSource.getOptions.find(o => o.value === getIn(form.values, name))}
          name={name}
          isClearable={isClearable}
          components={{ SelectContainer }}
          onChange={value => this.handleChange(form, value)}
          onBlur={() => form.setFieldTouched(name, true)}
          options={dataSource.getOptions}
          classNamePrefix="react-select"
          {...props}
        />
      );
    }
    return (
      <AsyncSelect
        value={this.state.value}
        name={name}
        isClearable={isClearable}
        onChange={value => this.handleChange(form, value)}
        onBlur={() => form.setFieldTouched(name, true)}
        defaultOptions
        cacheOptions
        components={{ SelectContainer }}
        loadOptions={inputValue => this.loadOptions(form, inputValue)}
        classNamePrefix="react-select"
        {...props}
      />
    );
  }
}

DropDown.propTypes = {
  ...Field.propTypes,
  dataSource: PropTypes.shape({
    getOptions: PropTypes.oneOfType([PropTypes.func, PropTypes.array]).isRequired,
  }).isRequired,
  handleChange: PropTypes.func,
  isClearable:  PropTypes.bool,
};

DropDown.defaultProps = {
  handleChange() {},
  isClearable: true,
};

export default DropDown;
