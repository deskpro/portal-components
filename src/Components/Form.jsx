import React from 'react';
import { Form as FormikForm } from 'formik';

const Form = ({
  ...props
}) => {
  return <FormikForm {...props} />
};

export default Form;