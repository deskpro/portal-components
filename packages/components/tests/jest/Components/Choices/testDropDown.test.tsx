import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { withFormik } from 'formik';
import { describe, expect, it, jest } from '@jest/globals';
import Form from '../../../../src/Components/Form';
import DropDown from '../../../../src/Components/Choices/DropDown';

jest.mock("react-select", () => {
  return {
    default: ({...props}) => <div>{JSON.stringify(props)}</div>
  }
});

describe('>>> DropDown --- Snapshot', () => {
  it('+++capturing Snapshot of DropDown', () => {
    const options = [
      { value: 'first', label: 'First' },
      { value: 'second', label: 'Second' },
      { value: 'last', label: 'Last' },
    ];
    const App = () => (
      <Form>
        <DropDown name="dropDown" dataSource={{ getOptions: options }} />
      </Form>
    );
    const FormikApp = withFormik({
      mapPropsToValues({
        dropDown
      }) {
        return {
          dropDown: dropDown || 'first'
        };
      },
      handleSubmit(values, { setSubmitting }) {
        setSubmitting(false);
      }
    })(App);
    const renderedValue = renderer.create(<FormikApp />).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});
