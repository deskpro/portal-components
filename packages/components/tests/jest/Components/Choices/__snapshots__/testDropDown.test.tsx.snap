// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> DropDown --- Snapshot +++capturing Snapshot of DropDown 1`] = `
<form
  action="#"
  className="dp-pc_form show-hover"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="dp-pc_field"
  >
    <label
      className="dp-pc_label"
      htmlFor="field_1"
    >
      
    </label>
    <div>
      {"styles":{},"value":{"value":"first","label":"First"},"name":"dropDown","isClearable":false,"isSearchable":true,"components":{},"menuPlacement":"auto","menuIsOpen":false,"options":[{"value":"first","label":"First"},{"value":"second","label":"Second"},{"value":"last","label":"Last"}],"closeMenuOnScroll":false,"classNamePrefix":"react-select","placeholder":"Select","label":"","errorsName":"","description":"","className":"","fClassName":"","id":null,"children":null,"required":false,"i18n":{},"closeOnBlur":true}
    </div>
  </div>
</form>
`;
